azure: Add the correct JSON tag to GatewayIP field in CiliumNode CRD

AnishShah opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

The incorrect JSON tag "-" was added to GatewayIP field in commit 8de6dc5
that was causing GatewayIP field to not get populated in CiliumNode CRD.
Since GatewayIP field was not populated, Cilium CNI was creating incorrect 
`ip rules and routes` that was affecting pod-to-pod connectivity.

```release-note
Fix a bug that was causing Azure IPAM with multiple pod subnets to not work.
```

AnishShah committed:
azure: Add the correct JSON tag to GatewayIP field in CiliumNode CRD

The incorrect JSON tag "-" was added to GatewayIP field in commit 8de6dc5
that was causing GatewayIP field to not get populated in CiliumNode CRD.

Signed-off-by: Anish Shah <anishshah@google.com>

aanm commented on pkg/azure/types/types.go:
```diff
@@ -113,7 +113,7 @@ type AzureInterface struct {
 	//
 	// +optional
-	GatewayIP string `json:"-"`
+	GatewayIP string `json:"gateway-ip"`
```
https://github.com/cilium/cilium/commit/8de6dc5a02b1af20d95bcc9c3783c9b08b2157c0 did break the API by adding the `json:"-"` tag.

Unfortunately, to not break the API, 
```
GatewayIP string `json:"gateway-ip"`
```
needs to be
```
GatewayIP string `json:"GatewayIP"`
```
as it was introduced in https://github.com/cilium/cilium/pull/11268/files#diff-54f340ec296f0270b40d38b4245bbb130a622e4b05315285f98b29629a4e0f31R116

Perhaps it is better to keep the `json:"GatewayIP"`, backport it to 1.9 and in 1.10 add a new field (`json:"gateway-ip"`) that would allow us to have consistency across all json tags in `AzureInterface` while deprecating `json:"GatewayIP"` and removing it in 1.11.

AnishShah replied:
@aanm, I don't understand how my change will break the API because it was introduced in https://github.com/cilium/cilium/pull/11268/files#diff-54f340ec296f0270b40d38b4245bbb130a622e4b05315285f98b29629a4e0f31R116. I'm adding a json tag whereas the code you're referring to seems to be using the struct field.

If I understand the codebase correctly, then this is what is happening - 
1. `cilium-operator-azure` will update the `status` of `CiliumNode` CR with `GatewayIP` information [here](https://github.com/cilium/cilium/blob/769123365c08746eb228fbd231b0cc1833eacca0/pkg/ipam/node.go#L784). But I think kube-apiserver ignores it as `GatewayIP` is not present in the CRD schema. 
2. `cilium-agent` tries to fetch `GatewayIP` of the AzureInterface from CiliumNode CR but since it is not present in the CRD schema, it will be empty. 

With my change which adds a `gateway-ip` field in CRD schema,
1. `cilium-operator-azure` will be successful in populating `gateway-ip` field in `CiliumNode` CR.
2. `cilium-agent` will be successful in fetching the `GatewayIP` of the AzureInterface as it is populated in `CiliumNode` CR.

I tested my changes in 1.18 Azure cluster and it seems to be working. I ensured that there is per-NIC routing table using this command `ip rule`. There was no routing table for the secondary NIC before my changes.

aanm replied:
@AnishShah when the field was introduced in https://github.com/cilium/cilium/pull/11268/files#diff-54f340ec296f0270b40d38b4245bbb130a622e4b05315285f98b29629a4e0f31R116 the JSON key of that field was `GatewayIP`. This PR is changing the API form `GatewayIP` to `gateway-ip`.

AnishShah replied:
Ok. Since this is a pull request against master, I'll add one commit which changes the JSON tag to `GatewayIP`. Another commit which deprecates this field and adds another field. Any idea what should I name the new field? Just `Gateway` or `GatewayAddress`

During backport to 1.9, we can just include the first commit.

AnishShah replied:
Also I'm not sure how I can make a struct field as deprecated. Any examples on how to do it?

christarazi replied:
We usually just mark the field as deprecated by putting a comment on the field saying so: https://github.com/cilium/cilium/blob/ffd02dd37aebbea366df9cadc752fe95fb2ba137/pkg/aws/eni/types/types.go#L49

AnishShah replied:
Fixed.

ti-mo commented on examples/crds/ciliumnodes.yaml:
```diff
@@ -253,6 +253,11 @@ spec:
                     items:
                       description: AzureInterface represents an Azure Interface
                       properties:
+                        GatewayIP:
```
Any particular reason why this field is capitalized?

AnishShah replied:
@ti-mo please refer to the above discussion with @aanm. This property-name is generated from JSON tag in `AzureInterface` struct in pkg/azure/types/types.go. We need to keep this capitalized so that it doesn't break the API.

ti-mo commented on examples/crds/ciliumnodes.yaml:
```diff
@@ -253,6 +253,11 @@ spec:
                       description: AzureInterface represents an Azure Interface
                       properties:
+                        GatewayIP:
+                          description: "GatewayIP is the interface subnet's default
```
Had to read this a few times; did you mean `the interface's subnet's default route`?

AnishShah replied:
Fixed.

ti-mo commented on examples/crds/ciliumnodes.yaml:
```diff
@@ -253,6 +253,11 @@ spec:
+                        GatewayIP:
+                          description: "GatewayIP is the interface subnet's default
+                            route \n OBSOLETE: This field is obsolete, please use
+                            Gateway field instead."
```
I'm not aware of where this docstring is used exactly, but does `Gateway` need to be capitalized here? Could it be turned into a fully-qualified name like `items.gateway` instead? (again, not sure about this)

AnishShah replied:
This is generated from docstring of `Gateway` field in `AzureInterface` struct in `pkg/azure/types/types.go` file. If you notice other fields, they too use capitalized field name in their docstring too.

ti-mo commented on examples/crds/ciliumnodes.yaml:
```diff
@@ -273,6 +278,9 @@ spec:
                             type: object
                           type: array
+                        gateway:
+                          description: Gateway is the interface subnet's default route
```
See above comment.

AnishShah replied:
same as above.

AnishShah committed:
azure: Add Gateway field to AzureInterface and deprecate GatewayIP field

The JSON tag for GatewayIP field does not follow the same convention as
other field. Inorder to have the same consistency across all JSON tags,
we introduce Gateway field and deprecate GatewayIP field.

Signed-off-by: Anish Shah <anishshah@google.com>

ti-mo reviewed (approved):
Thanks!
